pkgs/record_use/README.md

```diff
@@ -0,0 +1,96 @@
+# `package:record_use`
+
+
+> [!CAUTION]
+> This is an experimental package, and it's API can break at any time. Use at
+> your own discretion.
+
+This package provides the data classes for the usage recording feature in the
+Dart SDK.
+
+Dart objects with the `@RecordUse` annotation are being recorded at compile 
+time, providing the user with information. The information depends on the object
+being recorded.
+
+- If placed on a static method, the annotation means that arguments passed to
+the method will be recorded, as far as they can be inferred at compile time.
+- If placed on a class with a constant constructor, the annotation means that
+any constant instance of the class will be recorded. This is particularly useful
+when using the class as an annotation.
+
+## Example
+
+```dart
+import 'package:meta/meta.dart' show RecordUse;
+
+void main() {
+  print(SomeClass.stringMetadata(42));
+  print(SomeClass.doubleMetadata(42));
+  print(SomeClass.intMetadata(42));
+  print(SomeClass.boolMetadata(42));
+}
+
+class SomeClass {
+  @RecordMetadata('leroyjenkins')
+  @RecordUse()
+  static stringMetadata(int i) {
+    return i + 1;
+  }
+
+  @RecordMetadata(3.14)
+  @RecordUse()
+  static doubleMetadata(int i) {
+    return i + 1;
+  }
+
+  @RecordMetadata(42)
+  @RecordUse()
+  static intMetadata(int i) {
+    return i + 1;
+  }
+
+  @RecordMetadata(true)
+  @RecordUse()
+  static boolMetadata(int i) {
+    return i + 1;
+  }
+}
+
+@RecordUse()
```

Comment:
As mentioned in previous discussions, I strongly prefer us to distinguish what we're recording - instead of a general `record this` without saying what exactly should be recorded.

So I'd make this e.g. instead
```dart
@RecordConstantUse()
class Foo {
  final Object fieldA;
  const Foo(this.fieldA);
}
```
It tells our compiler to only track constants of this class and maybe its uses (and doesn't need to track anything else related to this class - e.g. class hierarchy, ...)

We can do it in a follow up CL, under the condition that this won't be put under the carpet and forgotten!

Reply:
What name the annotation has is entirely up to `package:meta` and the associated retrieval in the SDK - `record_use` has no knowledge of this apart from documentation examples. I am not opposed at all to having different names for different recordings.